spring-cloud-spanner-spring-data-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/springdata/it/AbstractBaseSpannerR2dbcIntegrationTest.java

```diff
@@ -0,0 +1,77 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner.r2dbc.springdata.it;
+
+import com.google.cloud.ServiceOptions;
+import com.google.cloud.spanner.Spanner;
+import com.google.cloud.spanner.SpannerOptions;
+import com.google.cloud.spanner.r2dbc.springdata.SpannerR2dbcDialect;
+import java.util.List;
+import java.util.UUID;
+import org.junit.jupiter.api.AfterEach;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.context.annotation.UserConfigurations;
+import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+
+/**
+ * Abstract base for Spanner R2DBC integration tests of {@link SpannerR2dbcDialect}.
+ */
+abstract class AbstractBaseSpannerR2dbcIntegrationTest {
+  private static final Logger log =
+      LoggerFactory.getLogger(AbstractBaseSpannerR2dbcIntegrationTest.class);
+  protected static final String PROJECT_NAME =
+      System.getProperty("gcp.project", ServiceOptions.getDefaultProjectId());
+  protected static final String DRIVER_NAME = "spanner";
+  protected static final String TEST_INSTANCE =
+      System.getProperty("spanner.instance", "reactivetest");
+  protected String testDatabase;
+  protected Spanner spanner;
+  protected ApplicationContextRunner contextRunner;
+
+  /** Initializes the integration test environment for the Spanner R2DBC dialect. */
+  void initializeTestEnvironment(Class configurationClass, String tableDdl) {
+    this.testDatabase = "testdb_" + UUID.randomUUID().toString().replace('-', '_').substring(0, 20);
+
+    this.spanner = SpannerOptions.newBuilder().setProjectId(PROJECT_NAME).build().getService();
+
+    if (configurationClass != null) {
+      this.contextRunner =
+          new ApplicationContextRunner()
+              .withPropertyValues("testDatabase=" + testDatabase)
+              .withConfiguration(
+                  UserConfigurations.of(
+                      configurationClass));
+    }
+
+    log.info("Creating database {}", testDatabase);
+    spanner.getDatabaseAdminClient().createDatabase(
+        TEST_INSTANCE, testDatabase, List.of(tableDdl));
+    log.info("Done creating database {}", testDatabase);
+  }
+
+  void initializeTestEnvironment(String tableDdl) {
+    initializeTestEnvironment(null, tableDdl);
+  }
+
+  @AfterEach
+  void cleanupTestInstanceAfterTest() {
+    log.info("Deleting database {}", testDatabase);
+    this.spanner.getDatabaseAdminClient().dropDatabase(TEST_INSTANCE, testDatabase);
```

Comment:
If something goes wrong and this cleanup isn't able to run, we'll have an orphaned database that never gets removed. Example of resolving this situation in google-cloud-java, by checking at the beginning of a test whether any stale resources exist from prior runs: https://github.com/googleapis/google-cloud-java/blob/main/java-compute/google-cloud-compute/src/test/java/com/google/cloud/compute/v1/integration/Util.java#L41-L57

Reply:
`@AfterEach` is guaranteed to run even if there is a test failure. Unlike Compute instances, Spanner databases don't incur additional costs, but there is a limit of 100 per instance. I would rather hold off the proactive additional cleanup that will slow down the tests, until it's clear that we need it.

Reply:
@AfterEach doesn't run if the machine fails - the exact problem we had with google-cloud-java. I leave it up to you.